File: language/runtime/android/kotlin/src/test/kotlin/dev/destack/runtime/android/module/permission/PermissionFixtures.kt
package dev.destack.runtime.android

import dev.destack.runtime.android.module.permission.PermissionEvents
import dev.destack.runtime.android.module.permission.PermissionRequests
import dev.destack.runtime.android.module.permission.RuntimeHostPermissionEvent
import dev.destack.runtime.android.module.permission.RuntimeHostPermissionRequest

/**
 * One recording permission request handler for Android host tests.
 */
internal class RecordingPermissionRequestHandler : PermissionRequests {
    val requests: MutableList<RuntimeHostPermissionRequest> = mutableListOf()
    var openSettingsCalls: Int = 0
    var openSettingsStatus: Int = 0
    var requestStatus: Int = 0

    override fun request(
        request: RuntimeHostPermissionRequest,
    ): Int {
        requests += request

        return requestStatus
    }

    override fun openSettings(): Int {
        openSettingsCalls += 1

        return openSettingsStatus
    }
}

/**
 * One recording permission event sink for Android host tests.
 */
internal class RecordingPermissionEventSink : PermissionEvents {
    val events: MutableList<RuntimeHostPermissionEvent> = mutableListOf()
    val callback: PermissionEvents = this

    override fun notifyPermissionResult(
        event: RuntimeHostPermissionEvent,
    ) {
        events += event
    }
}

/**
 * One no-op permission request handler for Android host tests.
 */
internal class NoopPermissionRequestHandler : PermissionRequests {
    override fun request(
        request: RuntimeHostPermissionRequest,
    ): Int = 0

    override fun openSettings(): Int {
        return 1
    }
}

File: language/runtime/apple/core/Tests/RuntimeHostAppleCoreTests/Fixtures/PermissionFixtures.swift
import RuntimeHostAppleCore

/// One recording permission request handler for Apple core tests.
@MainActor
final class RecordingPermissionRequestHandler: PermissionRequests {
  var requests: [RuntimeHostPermissionRequest] = []
  var openSettingsCalls: Int = 0
  var openSettingsStatus: UInt32 = hostStatusOk

  func request(_ request: RuntimeHostPermissionRequest) -> UInt32 {
    requests.append(request)

    return hostStatusOk
  }

  func openSettings() -> UInt32 {
    openSettingsCalls += 1

    return openSettingsStatus
  }
}

/// One recording permission event sink for Apple core tests.
@MainActor
final class RecordingPermissionEventSink: PermissionEvents {
  var events: [RuntimeHostPermissionEvent] = []

  func notifyPermissionResult(_ event: RuntimeHostPermissionEvent) {
    events.append(event)
  }
}
